File: simulador_python/simulador.py
import json

class Task:
    def __init__(self, id, offset, computation_time):
        self.id = id
        self.offset = offset
        self.computation_time = computation_time
        self.completion = 0
        self.waiting = 0
        self.turnaround = 0

def load_tasks_from_json(path):
    with open(path, 'r') as f:
        data = json.load(f)
        tasks = []
        for i, task in enumerate(data['tasks']):
            try:
                tasks.append(Task(i,
                                  task["offset"],
                                  task["computation_time"]))
            except KeyError as e:
                raise ValueError(f"Erro na task {i}: campo vazio -> {e}")
        return tasks

def fcfs(tasks):
    tasks.sort(key=lambda task: task.offset)
    time = 0
    for task in tasks:
        if time < task.offset:
            time = task.offset
        task.waiting = time - task.offset
        time += task.computation_time
        task.completion = time
        task.turnaround = task.completion - task.offset

    print("\nFCFS Scheduling:")
    for task in tasks:
        print(f"P{task.id} WT={task.waiting} TAT={task.turnaround}")
    print_detailed_timeline(tasks, time)

def sjf_non_preemptive(tasks):
    n = len(tasks)
    complete = 0
    time = 0
    visited = [False] * n
    result = []

    while complete < n:
        idx = -1
        min_computation_time = float('inf')
        for i, task in enumerate(tasks):
            if task.offset <= time and not visited[i] and task.computation_time < min_computation_time:
                min_computation_time = task.computation_time
                idx = i

        if idx == -1:
            time += 1
            continue

        task = tasks[idx]
        visited[idx] = True
        task.waiting = time - task.offset
        time += task.computation_time
        task.completion = time
        task.turnaround = task.completion - task.offset
        result.append(task)
        complete += 1

    print("\nSJF Non-Preemptive Scheduling:")
    for task in result:
        print(f"P{task.id} WT={task.waiting} TAT={task.turnaround}")
    print_detailed_timeline(result, time)

def print_detailed_timeline(tasks, total_time):
    print("\nTimeline detalhada:")
    for task in tasks:
        line = []
        start = task.completion - task.computation_time
        for t in range(total_time):
            if t >= task.offset and t < start:
                line.append('-')
            elif t >= start and t < task.completion:
                line.append('#')
            else:
                line.append('_')
        print(f"P{task.id}: {''.join(line)}")

if __name__ == "__main__":
    procs = load_tasks_from_json('simulador_python/file.json')
    fcfs([Task(task.id, task.offset, task.computation_time) for task in procs])
    sjf_non_preemptive([Task(task.id, task.offset, task.computation_time) for task in procs])

File: simulador_c/simulador.c
#include <stdio.h>
#include <stdlib.h>

typedef struct {
    int pid, arrival, burst, waiting, turnaround, completion, done;
} Process;

int load_processes(const char* filename, Process* p, int max) {
    FILE *file = fopen(filename, "r");
    if (!file) {
        printf("Erro ao abrir o arquivo\n");
        return 0;
    }

    int count = 0;
    while (fscanf(file, "%d %d %d", &p[count].pid, &p[count].arrival, &p[count].burst) == 3 && count < max) {
        p[count].done = 0;
        count++;
    }

    fclose(file);
    return count;
}

void fcfs(Process p[], int n) {
    int time = 0;
    for (int i = 0; i < n; i++) {
        if (time < p[i].arrival)
            time = p[i].arrival;
        p[i].waiting = time - p[i].arrival;
        time += p[i].burst;
        p[i].completion = time;
        p[i].turnaround = p[i].completion - p[i].arrival;
    }

    printf("\nFCFS Scheduling:\n");
    for (int i = 0; i < n; i++)
        printf("P%d WT=%d TAT=%d\n", p[i].pid, p[i].waiting, p[i].turnaround);
    print_detailed_timeline(p, n, time);
}

void sjf(Process p[], int n) {
    int complete = 0, time = 0;
    while (complete < n) {
        int idx = -1, min = 9999;
        for (int i = 0; i < n; i++) {
            if (!p[i].done && p[i].arrival <= time && p[i].burst < min) {
                min = p[i].burst;
                idx = i;
            }
        }

        if (idx == -1) {
            time++;
            continue;
        }

        p[idx].waiting = time - p[idx].arrival;
        time += p[idx].burst;
        p[idx].completion = time;
        p[idx].turnaround = p[idx].completion - p[idx].arrival;
        p[idx].done = 1;
        complete++;
    }

    printf("\nSJF Non-Preemptive Scheduling:\n");
    for (int i = 0; i < n; i++)
        printf("P%d WT=%d TAT=%d\n", p[i].pid, p[i].waiting, p[i].turnaround);
    print_detailed_timeline(p, n, time);
}

void print_detailed_timeline(Process p[], int n, int total_time) {
    printf("\nTimeline detalhada:\n");
    for (int i = 0; i < n; i++) {
        int start = p[i].completion - p[i].burst;
        printf("P%d: ", p[i].pid);
        for (int t = 0; t < total_time; t++) {
            if (t >= p[i].arrival && t < start)
                printf("-");
            else if (t >= start && t < p[i].completion)
                printf("#");
            else
                printf("_");
        }
        printf("\n");
    }
}

int main() {
    Process p[100];
    int n = load_processes("processos.txt", p, 100);

    fcfs(p, n);
    sjf(p, n);

    return 0;
}
